### DISCUSSION TITLE
Purpose of token in RTU client.

### DISCUSSION THREAD

**Discussion by EmbeddedDevver** [eModbus/eModbus#126] [Q&A] [answered]
Could anyone explain me the purpose of the token? 

e.g.
``` err = MBclient.addRequest(token, modbus_RemoteId, WRITE_HOLD_REGISTER, 10 - 1, targetValue);```

1) Does it mind if requests have the same token and if so, what can happen / can go wrong?
2) if a requests is transmitted but no (valid) response was given from the slave, does the driver automatically re-requests until a valid response was received? 

BC I now see, that if the slave gots disconnected, that the queue buffer gets full because no response is received from that slave.

I prefer to add a request, and then get a error (like timeout) or receive the data and stop after that. Now it seems the driver continues requesting data for as long it does not receive a valid reply. This causes the transmit buffer to reach it maximum requests (default = 100). And then also other nodes that are still available in the bus, cannot be serviced anymore as the transmit buffer is already full.

**Comment by Miq1** [comment, marked as answer]
1.: you will not be able to tell the responses one from the other in the ``onDataHandler`` and ``onErrorHandler`` callbacks without a different token, since the response is received asynchronously.
2.: a missing response will end in a E0 TIMEOUT error reported, all malformed responses will have error responses accordingly.

If you keep on sending requests without checking the responses, you may end up with a full request queue eventually.

You will need to write your ``onErrorHandler`` as well to cope with these situations...

**Reply by EmbeddedDevver** [reply]
Thank you for the answer. Very useful for me.

I have both handlers active (onDataHandler and onErroHandler).
Indeed for the slave that goes offline, I see that  I get a timeout in the error routine. 

It just came to my mind, that the default timeout is set to 

```#define DEFAULTTIMEOUT 2000 ```

I ask every slave for data at an interval of 5 seconds. But for the offline slave, I have 10 requests, and now I see that this timeout adds up to every requests, causing a delay of 2000mS times 10 = 20seconds. 

I need to solve this by disabling the slave when I see it's offline.

Thanks again.

**Comment by Miq1** [comment]
This is an hen-and-egg problem, I am afraid. You will not know the server is disconnected unless you will wait for some time for the response. There are slow RTU servers that will need some time to react. The Modbus RTU rules are requiring all else traffic to stop until either the server has responded or the client will send another request.

The only viable solution I see for your problem is to send a request and patiently wait some time. Note: you can set your own timeout for the connection that may be shorter than 2 seconds.

If all else fails, we may consider to add a kind of ``dropRequestsForServer()`` service call that will scan the queue and remove requests for that very server ID.
